infinite effects

remminiscent opened the pull request:
Implements Infinite effects.
### Related issues & PRs
* Fixes https://github.com/pmmp/PocketMine-MP/issues/6862
## Changes
### API changes
`EffectInstance->isInfinite()`
## Backwards compatibility
no break.... hopefully
## Tests
https://streamable.com/ffaecx

remminiscent committed:
infinite effects

remminiscent committed:
Merge branch 'minor-next' into infinite-effects

remminiscent commented:
pocketmine commands are so outrageously bad i dont even know how to add the infinite argument, im going to need your help to implement the command part @dktapps

remminiscent committed:
command part done

remminiscent commented:
how can i add the new message?

remminiscent commented:
ahhh, checking with tickable effects, they dont work... apologies.

what did you mean by "Clock" @dktapps

remminiscent committed:
Merge branch 'minor-next' into infinite-effects

dktapps commented:
> ahhh, checking with tickable effects, they dont work... apologies.
> 
> what did you mean by "Clock" @dktapps

I mean that the tickable effects need some way to know when to tick, currently it uses the duration, but if the duration is not updated that's obviously not gonna work.

Possibly we could just init them with `PHP_INT_MAX` and then treat anything with duration > `Limits::INT32_MAX` as infinite. Since it would take longer than the current age of the universe for the ticker to count down from `PHP_INT_MAX` to `Limits::INT32_MAX` anyway, that would allow infinite effects to still tick

dktapps commented on src/command/defaults/EffectCommand.php:
```diff
@@ -113,7 +117,14 @@ public function execute(CommandSender $sender, string $commandLabel, array $args
 			$effectManager->add($instance);
-			self::broadcastCommandMessage($sender, KnownTranslationFactory::commands_effect_success($effect->getName(), (string) $instance->getAmplifier(), $player->getDisplayName(), (string) ($instance->getDuration() / 20)));
+
+			$durationDisplay = $duration === Limits::INT32_MAX ? "infinite" : (string) ($instance->getDuration() / 20);
```
This will look weird for non-English users. I'm fairly sure there should be a dedicated string for this in vanilla if you check their language files.

remminiscent replied:
there is, however im unknowledgeable regarding language files and pocketmine translation stuff.

dktapps replied:
Just add a string to `resources/translations/eng.ini` following the PM format, and then run `composer update-codegen` to regenerate the `KnownTranslationFactory`

dktapps replied:
https://github.com/pmmp/PocketMine-MP/blob/stable/resources/translations/README.md

remminiscent committed:
Merge branch 'minor-next' into infinite-effects

remminiscent committed:
message

remminiscent committed:
Merge branch 'infinite-effects' of github-remminiscent:remminiscent/PocketMine-MP into infinite-effects

remminiscent commented:
ok ill take a go at what you said. 

> > ahhh, checking with tickable effects, they dont work... apologies.
> > what did you mean by "Clock" @dktapps
> 
> I mean that the tickable effects need some way to know when to tick, currently it uses the duration, but if the duration is not updated that's obviously not gonna work.
> 
> Possibly we could just init them with `PHP_INT_MAX` and then treat anything with duration > `Limits::INT32_MAX` as infinite. Since it would take longer than the current age of the universe for the ticker to count down from `PHP_INT_MAX` to `Limits::INT32_MAX` anyway, that would allow infinite effects to still tick

wouldnt an exception be sent?

`		
if($duration < 0 || $duration > Limits::INT32_MAX){
			throw new \InvalidArgumentException("Effect duration must be in range 0 - " . Limits::INT32_MAX . ", got $duration");
		}
`
		(idk how to reference a piece of code.)

dktapps commented:
> wouldnt an exception be sent?
> 
> ` if($duration < 0 || $duration > Limits::INT32_MAX){ throw new \InvalidArgumentException("Effect duration must be in range 0 - " . Limits::INT32_MAX . ", got $duration"); } ` (idk how to reference a piece of code.)

We can change that code if need be. I'm not too sure what implications there might be for plugins saving the duration into persistent storage though.

remminiscent commented:
im not gonna lie to you bro. i cant figure it out... sounds simple but i cant seem to know what the goal is

dktapps commented:
My thinking is that if we set the duration to a sufficiently large number, we can let it count down (so still allowing the ticking effects to work) while making sure its duration is so large that it'll never run out during a normal server runtime.
Then we can just check if `duration > Limits::INT32_MAX` to see if the effect is infinite.

I'm not sure how well this would work in practice though. It raises some questions about what should happen when doing `setDuration(INT32_MAX < number < PHP_INT_MAX)`. Should doing `setDuration(PHP_INT_MAX - 10)` be allowed?
- It would turn into `PHP_INT_MAX` after saving and reloading, since we can't save a 64-bit value for the duration. This might be confusing in itself.
- But if we don't allow it, then `getDuration()` would be able to return values that `setDuration()` doesn't accept (it would have to, so that code like [this](https://github.com/pmmp/PocketMine-MP/blob/55c110e5c700c713120b6063b24d64da9128ed0b/src/entity/effect/PoisonEffect.php#L42) could work correctly). This would also be confusing for plugin developers.

I haven't given this much thought, so I'm not sure what other possible options there are. But if the duration isn't allowed to count down, then anything that uses the poison-style approach to tick effects will break. So we're kinda constrained with our options in 5.x.

In 6.0 we could do this any way we like, but I doubt anyone would want to wait for 6.0 to get this seemingly trivial feature.

remminiscent commented:
i mean, for now we can do the php int max, and int32 max approach, where if the duration is PHP_INT_MAX, itll still count down, then for save and loading nbt storage, if the duration is above int32 max, we just save -1, and when loading, we load php_int_max.


then we can specifiy in the doc block that ONLY PHP_INT_MAX will provide an infinite effect, anything below it will provide a normal effect, but if someone does PHP_INT_MAX - 10, itll get converted to INT32_MAX - 10?

dktapps commented:
Don't love that. `$newEffect->setDuration($infiniteEffect->getDuration())` would then become non-infinite.

remminiscent commented:
so whats the plan now? change target branch to major next and make it an actual good implementation of infinite effects, or figure a workaround for minor next? @dktapps

dktapps commented:
Workaround would be preferred as I said before

remminiscent commented:
i cant think of any right now, if someone else has an idea, please give it

dktapps committed:
Merge branch 'minor-next' of github.com:pmmp/PocketMine-MP into infinite-effects

dktapps committed:
Alternative solution for infinite effects

The problem is that we need to ensure an updating ticker for periodic effects to continue
working correctly, but those are typically implemented using the duration ticker.

We don't want to BC break anyone implementing custom effects in PM5, so this is the best
we can do for now.

I did consider a bitflag on the duration, but that would've been subjected to math
operations like the ones done by SplashPotion and AreaEffectCloud, which would've produced
bizarre results and potentially wiped out the infinite status of the effect. To avoid
complicating setDuration(), a separate property seemed like the next best idea.

In this implementation, durations will underflow back to Limits::INT32_MAX when they
would have expired, ensuring a continuous counter. This might cause unexpected behaviour
for consumers of the API, but I don't see another viable solution for PM5.

dktapps commented:
I worked this over some locally. I've pushed my solution in https://github.com/pmmp/PocketMine-MP/pull/6866/commits/30ac82bd31a35ce12c2c3e804cc1c69e246bb218. Let me know what you think.

It's not ideal, but for the scope of PM5 I don't see a better option. I'd had plans to rework EffectInstance in PM6 anyway

remminiscent commented:
yep, looks great for PM5.👍

though a rework for the effect system would be great too, but we'll leave that for PM6

dktapps reviewed (dismissed):
For PM5 I think this will do. I plan on reworking effect duration handling in PM6 anyway (see #3897) so any hacks won't survive too long.

Dasciam commented on src/entity/effect/EffectInstance.php:
```diff
@@ -76,11 +77,22 @@ public function setDuration(int $duration) : EffectInstance{
 	 */
 	public function decreaseDuration(int $ticks) : EffectInstance{
-		$this->duration = max(0, $this->duration - $ticks);
+		//Leave the flag set during subtraction - this prevents the
```
It seems this comment is not finished

dktapps replied:
Actually I meant to delete it, must've missed a line

dktapps committed:
Delete dead comment

dktapps reviewed (approved):
I think this is as good as it gets in PM5. In PM6 it's planned to do some breaks to effect durations anyway, so the hacks will be cleaned up then.

In the meantime plugins can start preparing for that by making use of #6881

ShockedPlot7560 reviewed (approved):
LGTM for pm5
